File: android/app/src/main/java/app/northax/domain/engine/StrengthEngine.kt
package app.northax.domain.engine

import app.northax.domain.model.DailyReadiness
import app.northax.domain.model.ExerciseSuggestion
import app.northax.domain.model.GarminActivity
import app.northax.domain.model.GarminActivityType
import app.northax.domain.model.MuscleGroup
import app.northax.domain.model.StrengthSession

/** Deterministic strength session generator — offline fallback (iOS StrengthEngine). */
object StrengthEngine {

    fun generateSession(
        muscleGroups: List<MuscleGroup>,
        readiness: DailyReadiness,
        recentActivities: List<GarminActivity>,
    ): StrengthSession {
        val intensity = intensityFor(readiness)
        val warnings = buildRecoveryWarnings(muscleGroups, recentActivities)
        val exercises = buildExercises(muscleGroups, intensity)
        return StrengthSession(
            muscleGroups = muscleGroups,
            title = buildTitle(muscleGroups),
            exercises = exercises,
            duration = estimateDuration(exercises),
            intensityLabel = intensity.label,
            rationale = buildRationale(muscleGroups, readiness, warnings.size),
            recoveryWarnings = warnings,
        )
    }

    // MARK: - Intensity levels

    enum class Intensity(val label: String) {
        Heavy("Heavy"), Moderate("Moderate"), Light("Light");

        val primarySets: Int get() = when (this) { Heavy -> 4; Moderate -> 3; Light -> 2 }
        val accessorySets: Int get() = maxOf(2, primarySets - 1)

        val primaryReps: String get() = when (this) { Heavy -> "5–7"; Moderate -> "8–12"; Light -> "15–20" }
        val accessoryReps: String get() = when (this) { Heavy -> "8–12"; Moderate -> "10–15"; Light -> "15–20" }
        val primaryRest: String get() = when (this) { Heavy -> "2–3 min"; Moderate -> "90 sec"; Light -> "60 sec" }
        val accessoryRest: String get() = when (this) { Heavy -> "90 sec"; Moderate -> "60 sec"; Light -> "45 sec" }
    }

    private fun intensityFor(readiness: DailyReadiness): Intensity = when (readiness.status) {
        DailyReadiness.Status.Peak, DailyReadiness.Status.High -> Intensity.Heavy
        DailyReadiness.Status.Moderate -> Intensity.Moderate
        DailyReadiness.Status.Low, DailyReadiness.Status.Rest -> Intensity.Light
    }

    // MARK: - Exercise database

    private data class Movement(val name: String, val isCompound: Boolean, val note: String? = null)

    private val db: Map<MuscleGroup, List<Movement>> = mapOf(
        MuscleGroup.Chest to listOf(
            Movement("Barbell Bench Press", true, "Control the descent, full range"),
            Movement("Incline Dumbbell Press", true),
            Movement("Cable Chest Fly", false, "Squeeze at midpoint"),
            Movement("Dips", false),
        ),
        MuscleGroup.Back to listOf(
            Movement("Pull-Ups", true, "Drive elbows down, not hands"),
            Movement("Barbell Row", true, "Chest stays up, hinge at hips"),
            Movement("Seated Cable Row", false),
            Movement("Lat Pulldown", false),
        ),
        MuscleGroup.Shoulders to listOf(
            Movement("Overhead Press", true, "Full lockout at top"),
            Movement("Lateral Raise", false, "Lead with elbow, not wrist"),
            Movement("Face Pull", false, "External rotation at end range"),
            Movement("Arnold Press", false),
        ),
        MuscleGroup.Biceps to listOf(
            Movement("Barbell Curl", true),
            Movement("Hammer Curl", false),
            Movement("Incline Dumbbell Curl", false, "Stretch at bottom"),
        ),
        MuscleGroup.Triceps to listOf(
            Movement("Skull Crushers", true, "Keep elbows fixed"),
            Movement("Close-Grip Bench Press", true),
            Movement("Cable Pushdown", false),
        ),
        MuscleGroup.Quads to listOf(
            Movement("Back Squat", true, "Break parallel if mobility allows"),
            Movement("Leg Press", false),
            Movement("Hack Squat", false),
            Movement("Walking Lunge", false),
        ),
        MuscleGroup.Hamstrings to listOf(
            Movement("Romanian Deadlift", true, "Maintain neutral spine throughout"),
            Movement("Leg Curl", false),
            Movement("Nordic Curl", false, "Progress slowly — high injury risk if rushed"),
            Movement("Good Morning", false),
        ),
        MuscleGroup.Glutes to listOf(
            Movement("Hip Thrust", true, "Full hip extension at top"),
            Movement("Bulgarian Split Squat", true),
            Movement("Cable Kickback", false),
        ),
        MuscleGroup.Calves to listOf(
            Movement("Standing Calf Raise", true, "Full stretch at bottom"),
            Movement("Seated Calf Raise", false),
        ),
        MuscleGroup.Core to listOf(
            Movement("Dead Bug", true, "Lower back stays flat throughout"),
            Movement("Plank", false),
            Movement("Russian Twist", false),
            Movement("Hanging Leg Raise", false),
            Movement("Cable Crunch", false),
        ),
    )

    /** Movement names for one muscle group — the pick list for the live logger. */
    fun movements(group: MuscleGroup): List<String> = (db[group] ?: emptyList()).map { it.name }

    private fun buildExercises(muscleGroups: List<MuscleGroup>, intensity: Intensity): List<ExerciseSuggestion> {
        val exercisesPerGroup = if (muscleGroups.size <= 2) 3 else 2
        val result = mutableListOf<ExerciseSuggestion>()
        for (group in muscleGroups) {
            val selected = (db[group] ?: emptyList()).take(exercisesPerGroup)
            selected.forEachIndexed { i, movement ->
                val isFirst = i == 0
                result.add(
                    ExerciseSuggestion(
                        name = movement.name,
                        muscleGroup = group,
                        sets = if (isFirst) intensity.primarySets else intensity.accessorySets,
                        repsRange = if (isFirst) intensity.primaryReps else intensity.accessoryReps,
                        rest = if (isFirst) intensity.primaryRest else intensity.accessoryRest,
                        notes = if (isFirst) movement.note else null,
                    )
                )
            }
        }
        return result
    }

    private fun estimateDuration(exercises: List<ExerciseSuggestion>): Int {
        val totalSets = exercises.sumOf { it.sets }
        return (10 + totalSets * 3).coerceIn(30, 90) // 10 min warmup + ~3 min/set
    }

    private fun buildTitle(muscleGroups: List<MuscleGroup>): String = when (muscleGroups.size) {
        0 -> "Gym Session"
        1 -> "${muscleGroups[0].raw} Day"
        2 -> "${muscleGroups[0].raw} + ${muscleGroups[1].raw}"
        else -> {
            val hasPush = muscleGroups.any { it in listOf(MuscleGroup.Chest, MuscleGroup.Shoulders, MuscleGroup.Triceps) }
            val hasPull = muscleGroups.any { it in listOf(MuscleGroup.Back, MuscleGroup.Biceps) }
            val hasLegs = muscleGroups.any { it in listOf(MuscleGroup.Quads, MuscleGroup.Hamstrings, MuscleGroup.Glutes, MuscleGroup.Calves) }
            when {
                hasPush && !hasPull && !hasLegs -> "Push Day"
                !hasPush && hasPull && !hasLegs -> "Pull Day"
                !hasPush && !hasPull && hasLegs -> "Leg Day"
                else -> "Full Body"
            }
        }
    }

    // MARK: - Recovery warnings

    private fun buildRecoveryWarnings(
        muscleGroups: List<MuscleGroup>,
        recentActivities: List<GarminActivity>,
    ): List<String> {
        val lastStrength = recentActivities
            .filter { it.type == GarminActivityType.StrengthTraining }
            .maxByOrNull { it.startTime } ?: return emptyList()

        val hoursAgo = lastStrength.hoursAgo
        return muscleGroups.mapNotNull { group ->
            if (hoursAgo >= group.recoveryHours) return@mapNotNull null
            val remaining = (group.recoveryHours - hoursAgo).toInt()
            "${group.raw} trained ${hoursAgo.toInt()}h ago — ~${remaining}h until fully recovered. Reduce volume on these movements."
        }
    }

    // MARK: - Rationale

    private fun buildRationale(
        muscleGroups: List<MuscleGroup>,
        readiness: DailyReadiness,
        warningCount: Int,
    ): String {
        val groups = muscleGroups.take(3).joinToString(", ") { it.raw }
        var text = when (readiness.status) {
            DailyReadiness.Status.Peak ->
                "Readiness is at ${readiness.score}/100 — an ideal window for heavy work. The $groups session is loaded for strength adaptation: compound lifts first, heavier weights, longer rest intervals."
            DailyReadiness.Status.High ->
                "Readiness at ${readiness.score}/100 supports solid strength work. Stick to your working weights and focus on controlled reps — no reason to max out today, but no reason to hold back either."
            DailyReadiness.Status.Moderate ->
                "With readiness at ${readiness.score}/100, the session is dialled back to moderate intensity. Prioritise technique and mind-muscle connection. The volume is enough to maintain strength without adding recovery debt."
            DailyReadiness.Status.Low, DailyReadiness.Status.Rest ->
                "Readiness is low (${readiness.score}/100). If you train at all, keep loads very light — this is maintenance work only. Mobility or a walk may be a better investment of today's energy."
        }
        if (warningCount > 0) {
            text += " Recovery warnings are noted above — treat those muscle groups with care."
        }
        return text
    }
}

File: ios/NorthAx/Engine/StrengthEngine.swift
import Foundation

struct StrengthEngine {

    // MARK: - Public entry point

    static func generateSession(
        muscleGroups: [MuscleGroup],
        readiness: DailyReadiness,
        recentActivities: [GarminActivity]
    ) -> StrengthSession {
        let intensity = intensityFor(readiness: readiness)
        let warnings  = buildRecoveryWarnings(for: muscleGroups, recentActivities: recentActivities)
        let exercises = buildExercises(for: muscleGroups, intensity: intensity)
        let duration  = estimateDuration(exercises: exercises)
        let title     = buildTitle(for: muscleGroups)
        let rationale = buildRationale(muscleGroups: muscleGroups, readiness: readiness,
                                        intensity: intensity, warningCount: warnings.count)

        return StrengthSession(
            muscleGroups: muscleGroups,
            title: title,
            exercises: exercises,
            duration: duration,
            intensityLabel: intensity.label,
            rationale: rationale,
            recoveryWarnings: warnings
        )
    }

    // MARK: - Intensity levels

    enum Intensity {
        case heavy, moderate, light

        var label: String {
            switch self {
            case .heavy:    return "Heavy"
            case .moderate: return "Moderate"
            case .light:    return "Light"
            }
        }

        var primarySets: Int  { switch self { case .heavy: return 4; case .moderate: return 3; case .light: return 2 } }
        var accessorySets: Int { Swift.max(2, primarySets - 1) }

        var primaryReps: String   { switch self { case .heavy: return "5–7"; case .moderate: return "8–12"; case .light: return "15–20" } }
        var accessoryReps: String { switch self { case .heavy: return "8–12"; case .moderate: return "10–15"; case .light: return "15–20" } }
        var primaryRest: String   { switch self { case .heavy: return "2–3 min"; case .moderate: return "90 sec"; case .light: return "60 sec" } }
        var accessoryRest: String { switch self { case .heavy: return "90 sec"; case .moderate: return "60 sec"; case .light: return "45 sec" } }
    }

    private static func intensityFor(readiness: DailyReadiness) -> Intensity {
        switch readiness.status {
        case .peak, .high:     return .heavy
        case .moderate:        return .moderate
        case .low, .rest:      return .light
        }
    }

    // MARK: - Exercise database

    private static let db: [MuscleGroup: [(name: String, isCompound: Bool, note: String?)]] = [
        .chest: [
            ("Barbell Bench Press",      true,  "Control the descent, full range"),
            ("Incline Dumbbell Press",   true,  nil),
            ("Cable Chest Fly",          false, "Squeeze at midpoint"),
            ("Dips",                     false, nil)
        ],
        .back: [
            ("Pull-Ups",                 true,  "Drive elbows down, not hands"),
            ("Barbell Row",              true,  "Chest stays up, hinge at hips"),
            ("Seated Cable Row",         false, nil),
            ("Lat Pulldown",             false, nil)
        ],
        .shoulders: [
            ("Overhead Press",           true,  "Full lockout at top"),
            ("Lateral Raise",            false, "Lead with elbow, not wrist"),
            ("Face Pull",                false, "External rotation at end range"),
            ("Arnold Press",             false, nil)
        ],
        .biceps: [
            ("Barbell Curl",             true,  nil),
            ("Hammer Curl",              false, nil),
            ("Incline Dumbbell Curl",    false, "Stretch at bottom")
        ],
        .triceps: [
            ("Skull Crushers",           true,  "Keep elbows fixed"),
            ("Close-Grip Bench Press",   true,  nil),
            ("Cable Pushdown",           false, nil)
        ],
        .quads: [
            ("Back Squat",               true,  "Break parallel if mobility allows"),
            ("Leg Press",                false, nil),
            ("Hack Squat",               false, nil),
            ("Walking Lunge",            false, nil)
        ],
        .hamstrings: [
            ("Romanian Deadlift",        true,  "Maintain neutral spine throughout"),
            ("Leg Curl",                 false, nil),
            ("Nordic Curl",              false, "Progress slowly — high injury risk if rushed"),
            ("Good Morning",             false, nil)
        ],
        .glutes: [
            ("Hip Thrust",               true,  "Full hip extension at top"),
            ("Bulgarian Split Squat",    true,  nil),
            ("Cable Kickback",           false, nil)
        ],
        .calves: [
            ("Standing Calf Raise",      true,  "Full stretch at bottom"),
            ("Seated Calf Raise",        false, nil)
        ],
        .core: [
            ("Dead Bug",                 true,  "Lower back stays flat throughout"),
            ("Plank",                    false, nil),
            ("Russian Twist",            false, nil),
            ("Hanging Leg Raise",        false, nil),
            ("Cable Crunch",             false, nil)
        ]
    ]

    /// Movement names for one muscle group — the pick list for the live logger.
    static func movements(for group: MuscleGroup) -> [String] {
        (db[group] ?? []).map(\.name)
    }

    private static func buildExercises(for muscleGroups: [MuscleGroup], intensity: Intensity) -> [ExerciseSuggestion] {
        let exercisesPerGroup = muscleGroups.count <= 2 ? 3 : 2
        var result: [ExerciseSuggestion] = []

        for group in muscleGroups {
            let movements = db[group] ?? []
            let selected  = movements.prefix(exercisesPerGroup)

            for (i, movement) in selected.enumerated() {
                let isFirst  = i == 0
                let sets     = isFirst ? intensity.primarySets   : intensity.accessorySets
                let reps     = isFirst ? intensity.primaryReps   : intensity.accessoryReps
                let rest     = isFirst ? intensity.primaryRest   : intensity.accessoryRest
                let note     = isFirst ? movement.note : nil
                result.append(ExerciseSuggestion(
                    name: movement.name, muscleGroup: group,
                    sets: sets, repsRange: reps, rest: rest, notes: note
                ))
            }
        }
        return result
    }

    private static func estimateDuration(exercises: [ExerciseSuggestion]) -> Int {
        let totalSets = exercises.reduce(0) { $0 + $1.sets }
        return min(90, max(30, 10 + totalSets * 3)) // 10 min warmup + ~3 min/set
    }

    private static func buildTitle(for muscleGroups: [MuscleGroup]) -> String {
        switch muscleGroups.count {
        case 0: return "Gym Session"
        case 1: return "\(muscleGroups[0].rawValue) Day"
        case 2: return "\(muscleGroups[0].rawValue) + \(muscleGroups[1].rawValue)"
        default:
            // Classify as Push / Pull / Legs / Full Body
            let hasPush = muscleGroups.contains(where: { [.chest, .shoulders, .triceps].contains($0) })
            let hasPull = muscleGroups.contains(where: { [.back, .biceps].contains($0) })
            let hasLegs = muscleGroups.contains(where: { [.quads, .hamstrings, .glutes, .calves].contains($0) })
            if hasPush && !hasPull && !hasLegs { return "Push Day" }
            if !hasPush && hasPull && !hasLegs { return "Pull Day" }
            if !hasPush && !hasPull && hasLegs { return "Leg Day" }
            return "Full Body"
        }
    }

    // MARK: - Recovery warnings

    private static func buildRecoveryWarnings(
        for muscleGroups: [MuscleGroup],
        recentActivities: [GarminActivity]
    ) -> [String] {
        let lastStrength = recentActivities
            .filter { $0.type == .strengthTraining }
            .min(by: { $0.startTime > $1.startTime })

        guard let last = lastStrength else { return [] }
        let hoursAgo = last.hoursAgo

        return muscleGroups.compactMap { group in
            guard hoursAgo < Double(group.recoveryHours) else { return nil }
            let remaining = Int(Double(group.recoveryHours) - hoursAgo)
            return "\(group.rawValue) trained \(Int(hoursAgo))h ago — ~\(remaining)h until fully recovered. Reduce volume on these movements."
        }
    }

    // MARK: - Rationale

    private static func buildRationale(
        muscleGroups: [MuscleGroup],
        readiness: DailyReadiness,
        intensity: Intensity,
        warningCount: Int
    ) -> String {
        let groups = muscleGroups.prefix(3).map(\.rawValue).joined(separator: ", ")

        var text: String
        switch readiness.status {
        case .peak:
            text = "Readiness is at \(readiness.score)/100 — an ideal window for heavy work. The \(groups) session is loaded for strength adaptation: compound lifts first, heavier weights, longer rest intervals."
        case .high:
            text = "Readiness at \(readiness.score)/100 supports solid strength work. Stick to your working weights and focus on controlled reps — no reason to max out today, but no reason to hold back either."
        case .moderate:
            text = "With readiness at \(readiness.score)/100, the session is dialled back to moderate intensity. Prioritise technique and mind-muscle connection. The volume is enough to maintain strength without adding recovery debt."
        case .low, .rest:
            text = "Readiness is low (\(readiness.score)/100). If you train at all, keep loads very light — this is maintenance work only. Mobility or a walk may be a better investment of today's energy."
        }

        if warningCount > 0 {
            text += " Recovery warnings are noted above — treat those muscle groups with care."
        }
        return text
    }
}
